### PR TITLE
Fix pyxdameraulevenshtein build system

### PR DESCRIPTION
**Pull request by onyb**
Fix `setup.py` which fails while trying to bootstrap `numpy`. Please activate TravisCI builds on this repository before merging this PR. See a test run on my fork: https://travis-ci.org/onyb/pyxDamerauLevenshtein/builds/145873118

 * Setup Travis CI builds
 * Updated build status in `README.md`
 * Implemented basic tests from examples dir

Signed-off-by: Anirudha Bose <ani07nov@gmail.com>

### CODE REVIEW

**Comment by gfairchild** [comment]
This is awesome. I've wanted to play around with GitHub's Travis integration for a while, but I just haven't had the time. Thanks for kickstarting this.

I'll review this PR more thoroughly in a few days once some stuff here at work dies down.

**Comment by onyb** [comment]
@gfairchild You might also want to update the `pyxdameraulevenshtein.c` file and cut a new PyPI release.

**Comment by gfairchild** [comment]
Yeah, I definitely will.

**Comment by onyb** [comment]
@gfairchild Any update on this?

**Comment by gfairchild** [comment]
I'm hoping to finally review the code this weekend. You submitted it at a particularly bad time for me - I've been out of town 4 weekends straight.

File: pyxdameraulevenshtein/pyxdameraulevenshtein.pyx [hunk 1]
```diff
@@ -23,7 +23,7 @@
 
 from cpython.version cimport PY_MAJOR_VERSION
 from libc.stdlib cimport calloc, free
-import numpy as np
```
Comment (gfairchild) [inline pyxdameraulevenshtein/pyxdameraulevenshtein.pyx]: Why did you remove this import? The [Cython docs](http://docs.cython.org/en/latest/src/tutorial/numpy.html#adding-types) both `import` and `cimport` numpy.

File: requirements.txt [hunk 2]
```diff
@@ -1,2 +1 @@
 Cython==0.24.1
-numpy==1.11.1
```
Comment (gfairchild) [inline requirements.txt]: Why are you removing this dependency? `requirements.txt` is only used when setting up a virtual environment, which will only happen when someone clones the repo and runs tests and whatnot. They'll need numpy for that. This is separate from the numpy dependency handling in `setup.py`.

Comment (onyb) [inline requirements.txt]: Running `python setup.py install` automatically installs `numpy`, as you can also see from the test and the `.travis.yml` file. I noticed that keeping `numpy` in the `requirements.txt` was not a necessity.

Comment (gfairchild) [inline requirements.txt]: I use `requirements.txt` when setting up a virtual environment, which I do every time I start a new Python project. I'd like to keep `requirement.txt` unaltered. It's separate from `setup.py` and `.travis.yml`. Basically, it lays out all of the project's dependencies, implicit *and* explicit.

File: pyxdameraulevenshtein/pyxdameraulevenshtein.pyx (hunk 1 above)
Comment (onyb) [inline pyxdameraulevenshtein/pyxdameraulevenshtein.pyx]: @gfairchild The `cimport` should have taken care of most of the `numpy` interfaces as exposed [here](https://github.com/cython/cython/blob/master/Cython/Includes/numpy/__init__.pxd). However, the `vectorize` function doesn't appear to be available, hence we'll need the `import` for that. I forgot to re-cythonize, hence missed this. I'll revert this hunk.

File: requirements.txt (hunk 2 above)
Comment (onyb) [inline requirements.txt]: @gfairchild That's fine. I'll revert this as well.

Comment (gfairchild) [inline requirements.txt]: Great, thanks!

File: pyxdameraulevenshtein/pyxdameraulevenshtein.pyx (hunk 1 above)
Comment (gfairchild) [inline pyxdameraulevenshtein/pyxdameraulevenshtein.pyx]: Cool. :) I'm about to enable Travis CI as well.

**Comment by gfairchild** [comment]
Ok, Travis CI is now enabled on this project. I *think* it'll get triggered once you commit those reverts.

**Comment by onyb** [comment]
@gfairchild All done. :)

File: setup.py [hunk 3]
```diff
@@ -17,28 +17,51 @@
...
+from setuptools.command.build_ext import build_ext as _build_ext
 
-metadata = dict(
-    name='pyxDamerauLevenshtein',
-    version='1.4.1',
-    description="""pyxDamerauLevenshtein implements the Damerau-Levenshtein (DL) edit distance algorithm for Python in Cython for high performance.""",
-    long_description="""pyxDamerauLevenshtein implements the Damerau-Levenshtein (DL) edit distance algorithm for Python in Cython for high performance. Courtesy `Wikipedia <http://en.wikipedia.org/wiki/Damerau%E2%80%93Levenshtein_distance>`_:
 
-In information theory and computer science, the Damerau-Levenshtein distance (named after Frederick J. Damerau and Vladimir I. Levenshtein) is a "distance" (string metric) between two strings, i.e., finite sequence of symbols, given by counting the minimum number of operations needed to transform one string into the other, where an operation is defined as an insertion, deletion, or substitution of a single character, or a transposition of two adjacent characters.
+class build_ext(_build_ext):
```
Comment (gfairchild) [inline setup.py]: I'm about ready to accept this PR, but looking over your changes one more time, I'm a little confused on what this wrapper class is for. Would you mind explaining a bit?

Comment (onyb) [inline setup.py]: It is a small hack to override the default `build_ext` of setuptools so as to not complain about missing numpy package. See this Stackoverflow answer for details: http://stackoverflow.com/a/21621689/1946230

I think the `build_ext` provided by Cython won't need this hack, but I guess you don't want to keep Cython as a hard dependency.

Comment (gfairchild) [inline setup.py]: Awesome, nice work! Accepting this PR!
